cider/exec/nextgen/operators/SourceNode.cpp

```diff
@@ -19,11 +19,46 @@
  * under the License.
  */
 #include "exec/nextgen/operators/SourceNode.h"
+#include "exec/module/batch/ArrowABI.h"
 
+#include "exec/nextgen/jitlib/base/JITFunction.h"
+#include "exec/nextgen/jitlib/base/ValueTypes.h"
 #include "exec/template/common/descriptors/InputDescriptors.h"
 #include "util/Logger.h"
 
 namespace cider::exec::nextgen::operators {
-SourceNode::SourceNode(const ExprPtrVector& input_cols)
-    : OpNode("SourceNode"), input_cols_(input_cols) {}
+
+void SourceTranslator::consume(Context& context) {
+  codegen(context);
+}
+
+void SourceTranslator::codegen(Context& context) {
+  auto func = context.query_func_;
+  auto inputs = node_.getExprs();
+  // get ArrowArray pointer
+  auto arrow_pointer = func->getArgument(0);
```

Comment:
not use `auto` too much, use explicit type will improve codes' readability, sometimes will eliminate potential bugs.

Reply:
Roger that,  in this scene just for simplified code, such as **node_.getExprs** 's return type will be very complex. I have changed some places

Reply:
> not use auto too much, use explicit type will improve codes' readability, sometimes will eliminate potential bugs.

maybe you should open `Inlay` in some IDE.